### ISSUE TITLE
Improve Blazor localization handling

### ISSUE THREAD

**Issue by PieterjanDeClippel** [dotnet/aspnetcore#38675]
## Summary

Building a localized website with Blazor Server is really inconvenient at the moment. I've consulted the documentation, tried multiple approaches but Blazor just doesn't seem to make it convenient.

## Motivation and goals

For SEO purposes it is recommended to have the language either as an URL segment or query parameter. I prefer using a `lang` query parameter which needs to be monitored, and the culture updated accordingly.

[Here is an example angular app](https://github.com/PieterjanDeClippel/AngularI18N) with the approach I'm currently using in other angular apps.

### Breakdown

In the AppComponent the query parameter is monitored and changes the language of the entire application:

```
constructor(private route: ActivatedRoute, private translateService: TranslateService) {
  const defaultLang = 'en';
  this.route.queryParams.subscribe((params) => {
    const lang = params['lang'];
    if (!!lang) {
      this.translateService.use(lang);
    } else {
      this.translateService.use(defaultLang);
    }
  });
}
```

At the client-side the `TranslateModule` services are provided through the following `module import`:

    TranslateModule.forRoot({
      loader: {
        provide: TranslateLoader,
        useFactory: (http: HttpClient) => new TranslateHttpLoader(http),
        deps: [ HttpClient ]
      }
    }),

During SSR the `TranslateModule` services are provided through the following `module import` in the app.browser.module.ts:

    TranslateModule.forRoot({
      loader: {
        provide: TranslateLoader,
        useFactory: () => new TranslateJsonLoader()
      }
    })

With the `TranslateJsonLoader` like this:

```
import { TranslateLoader } from '@ngx-translate/core';
import { of } from 'rxjs';
import * as translationEn from '../../assets/i18n/en.json';
import * as translationNl from '../../assets/i18n/nl.json';

export class TranslateJsonLoader implements TranslateLoader {
  constructor() {
  }

  public getTranslation(lang: string) {
    switch (lang) {
      case 'nl':
        return of(translationNl);
      default:
        return of(translationEn);
    }
  }
}
```

Now you can simply use the `TranslatePipe` on any component like a charm:

    {{ 'message' | translate }}

I'd expect something similarily simple to be possible in Blazor Server.
- Define your `resx` files with your translations
- Use `@translateService.Translate("message")` in your razor views
- Change the language using `@onclick="@translateService.SetLocale("en-EN")"`
- No dependency on a cookie to select the language, but have the language passed through the url

## In scope

1) How to specify your translations => most likely through `resx` files
2) Exposure of the language to the application
    a) Due to SEO considerations, the language should be present in the URL by preference
    b) Most likely some other developers would rather like to have the language in a cookie
3) How to change the language
    a) Could be through a `TranslateService`
    b) I myself just use `router.navigate([], { queryParams: { lang: 'en-EN' } });`

**Comment by pranavkm** [comment]
Blazor Server uses the request localization middleware for it's localization which supports query string among other ways to specify the current request culture. See https://docs.microsoft.com/aspnet/core/fundamentals/localization?view=aspnetcore-6.0#localization-middleware. I think the docs recommend cookies because it takes away the overhead of figuring out how to persist a user's preference it across multiple browser sessions.

> Use @translateService.Translate("message") in your razor views

Have you had a look at `IStringLocalizer`? It looks suspiciously similar to your translator APIs.

>  I myself just use router.navigate([], { queryParams: { lang: 'en-EN' } });

The equivalent is fairly similar in a Blazor Server once you've configured the query string provider: `NavigationManager.NavigateTo(NavigationManager.GetUriWithQueryParameter("culture", userSelectedCulture), forceReload: true);`

**Comment by PieterjanDeClippel** [comment]
Thanks for the response. I did look into this yesterday too, and didn't know how to get started with it rightaway. However I've created a [boilerplate with the code you provided](https://github.com/PieterjanDeClippel/BlazorI18N). Somehow I'm getting the same results as I'm experiencing with one of my previous approaches:

    <NavLink class="nav-link" href="@navigationManager.GetUriWithQueryParameter("culture", "en-EN")">
        English
    </NavLink>

The prerendered result is localized according to the query parameter provided, but after websassembly takes over, the views return to the default language. Now on top of this, since your suggestion uses `forceReload: true`, the page is refreshed. This is something I'd rather avoid.

(Do we continue this discussion on StackOverflow?)
